src/utilities.py

```diff
@@ -0,0 +1,183 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Utilities used by the charm."""
+
+import functools
+import logging
+import os
+import subprocess  # nosec B404
+import time
+from typing import Callable, Optional, Sequence, Type, TypeVar
+
+from typing_extensions import ParamSpec
+
+from errors import SubprocessError
+
+logger = logging.getLogger(__name__)
+
+
+# Parameters of the function decorated with retry
+ParamT = ParamSpec("ParamT")
+# Return type of the function decorated with retry
+ReturnT = TypeVar("ReturnT")
+
+
+# This decorator has default arguments, one extra argument is not a problem.
+def retry(  # pylint: disable=too-many-arguments
+    exception: Type[Exception] = Exception,
+    tries: int = 1,
+    delay: float = 0,
+    max_delay: Optional[float] = None,
+    backoff: float = 1,
+    local_logger: logging.Logger = logger,
+) -> Callable[[Callable[ParamT, ReturnT]], Callable[ParamT, ReturnT]]:
+    """Parameterize the decorator for adding retry to functions.
+
+    Args:
+        exception: Exception type to be retried.
+        tries: Number of attempts at retry.
+        delay: Time in seconds to wait between retry.
+        max_delay: Max time in seconds to wait between retry.
+        backoff: Factor to increase the delay by each retry.
+        logger: Logger for logging.
+
+    Returns:
+        The function decorator for retry.
+    """
+
+    def retry_decorator(
+        func: Callable[ParamT, ReturnT],
+    ) -> Callable[ParamT, ReturnT]:
+        """Decorate function with retry.
+
+        Args:
+            fn: The function to decorate.
+
+        Returns:
+            The resulting function with retry added.
+        """
+
+        @functools.wraps(func)
+        def fn_with_retry(*args, **kwargs) -> ReturnT:
+            """Wrap the function with retries."""
+            remain_tries, current_delay = tries, delay
+
+            for _ in range(tries):
+                try:
+                    return func(*args, **kwargs)
+                # Error caught is set by the input of the function.
+                except exception as err:  # pylint: disable=broad-exception-caught
+                    remain_tries -= 1
+
+                    if remain_tries == 0:
+                        if local_logger is not None:
+                            local_logger.exception("Retry limit of %s exceed: %s", tries, err)
+                        raise
+
+                    if local_logger is not None:
```

Comment:
Shouldn't we force this logger to be defined?

Reply:
Yes, it might be a good idea.

Reply:
Minor comments